File: go/internal/smbagent/accounts.go
package smbagent

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"slices"

	"github.com/heavycaffeiner/stowcloud/go/internal/vfs"
	"sort"
	"strings"
)

// Reconciling the system account file and the credential database.
//
// The daemon resolves an SMB login to a system account by name, so every
// account named in a share needs an entry even though none of them ever owns a
// file. The server renders those entries; this puts them where the lookup
// looks, and imports the matching hashes.
//
// Rebuilt from scratch on every pass rather than diffed, which is what makes an
// account removed from the server's registry disappear here too instead of
// lingering as a stale entry forever.

// managedMarker is written into the comment field of every account this agent
// creates, and is the only thing that makes one eligible for removal. An
// account without it belongs to somebody else and is never touched.
const managedMarker = "sc-managed-smb"

// maxAccountName is the portable limit for a system account name.
const maxAccountName = 32

// Entry is one rendered account line, split far enough to reason about.
type Entry struct {
	Name string
	UID  string
	GID  string
	// Line is the whole line, marker already stamped.
	Line string
}

// ParseRendered reads what the server rendered.
//
// Anything that is not a seven-field line is dropped rather than passed
// through: this content ends up in the system account file, and a malformed
// line there breaks the name lookup for every account after it.
func ParseRendered(body string) []Entry {
	var out []Entry
	for _, line := range strings.Split(body, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		f := strings.Split(line, ":")
		if len(f) < 7 || f[0] == "" {
			continue
		}
		f[4] = managedMarker
		out = append(out, Entry{Name: f[0], UID: f[2], GID: f[3], Line: strings.Join(f, ":")})
	}
	return out
}

// ValidName holds a name to the portable rule before anything writes it into
// the system account file.
//
// The server's account names reach account-creation territory here. A name
// starting with a dash would also be argument injection into the credential
// tool.
func ValidName(name string) bool {
	if name == "" || len(name) > maxAccountName {
		return false
	}
	if first := name[0]; (first < 'a' || first > 'z') && first != '_' {
		return false
	}
	for i := 1; i < len(name); i++ {
		c := name[i]
		switch {
		case c >= 'a' && c <= 'z', c >= '0' && c <= '9', c == '_', c == '-':
		default:
			return false
		}
	}
	return true
}

// managed reports whether an account line is one of ours.
func managed(line string) bool {
	f := strings.Split(line, ":")
	return len(f) > 4 && f[4] == managedMarker
}

// Collisions refuses to sync at all when an SMB account collides with a
// pre-existing system account this agent did not create.
//
// Adopting one silently would give that account's name a credential, and
// removing it later would delete a system user. The numeric half is not
// cosmetic either: the import tool resolves a credential line to an account by
// number and takes whatever name the reverse lookup answers with, so a shared
// number attaches the credential to the wrong name.
func Collisions(desired []Entry, passwd string) []string {
	var out []string
	for _, e := range desired {
		if !ValidName(e.Name) {
			out = append(out, fmt.Sprintf("%q is not a valid system account name", e.Name))
			continue
		}
		for _, line := range strings.Split(passwd, "\n") {
			if line == "" || managed(line) {
				continue
			}
			f := strings.Split(line, ":")
			if len(f) < 3 {
				continue
			}
			name, uid := f[0], f[2]
			if name == e.Name {
				out = append(out, fmt.Sprintf("%q already exists as a system account this agent did not create", e.Name))
			}
			if uid == e.UID {
				out = append(out, fmt.Sprintf(
					"user id %s is already %q, an account this agent did not create: move the configured service id clear of the host's real users",
					e.UID, name))
			}
		}
	}
	sort.Strings(out)
	return slices.Compact(out)
}

// MissingGroups lists groups the rendered accounts reference that do not
// exist.
//
// This agent does not invent groups, and an account whose group resolves to
// nothing breaks the name lookup just as surely as a missing account.
func MissingGroups(desired []Entry, groupFile string) []string {
	have := map[string]bool{}
	for _, l := range strings.Split(groupFile, "\n") {
		if f := strings.Split(l, ":"); len(f) > 2 {
			have[f[2]] = true
		}
	}
	var out []string
	for _, e := range desired {
		if !have[e.GID] {
			out = append(out, e.GID)
		}
	}
	sort.Strings(out)
	return slices.Compact(out)
}

// Rebuild is the new account file: everything that is not ours, then ours.
func Rebuild(current string, desired []Entry) string {
	out := make([]byte, 0, len(current))
	for _, line := range strings.Split(current, "\n") {
		if line == "" || managed(line) {
			continue
		}
		out = append(out, line...)
		out = append(out, '\n')
	}
	for _, e := range desired {
		out = append(out, e.Line...)
		out = append(out, '\n')
	}
	return string(out)
}

// passwdMode is what the system account file has to be: every name lookup on
// the machine reads it, so one it cannot read breaks all of them.
const passwdMode = 0o644

// WritePasswd replaces the account file so a reader never sees a partial one,
// and so a machine that loses power mid-write comes back with one file or the
// other rather than a truncated roster.
//
// That still leaves the ordinary lost-update race against a concurrent account
// creation. The system's own lock is not reachable from here, and a file
// server's SMB roster is not a multi-writer account file.
func WritePasswd(path, body string) error {
	// The mode is applied to the staged file rather than left to the open,
	// because the process mask would narrow it.
	err := vfs.ReplaceFileDurable(path, passwdMode, func(f *os.File) error {
		_, werr := f.WriteString(body)
		return werr
	})
	if err != nil {
		return fmt.Errorf("replacing the account file: %w", err)
	}
	return nil
}

// Import loads the rendered hashes into the credential database.
func Import(smbpasswd, passdb string) error {
	out, err := exec.Command("pdbedit", //nolint:gosec // G204 flags a variable command: both arguments are the agent's own configured paths, and neither is caller data.
		"-i", "smbpasswd:"+filepath.Clean(smbpasswd),
		"-e", "tdbsam:"+filepath.Clean(passdb)).CombinedOutput()
	if err != nil {
		return fmt.Errorf("importing credentials: %w: %s", err, strings.TrimSpace(string(out)))
	}
	return nil
}

// PassdbNames lists the accounts the credential database currently holds.
func PassdbNames() ([]string, error) {
	out, err := exec.Command("pdbedit", "-L").Output()
	if err != nil {
		return nil, fmt.Errorf("listing credentials: %w", err)
	}
	var names []string
	for _, l := range strings.Split(string(out), "\n") {
		name, _, _ := strings.Cut(l, ":")
		if strings.TrimSpace(name) != "" {
			names = append(names, name)
		}
	}
	return names, nil
}

// Prune drops credentials for accounts that are no longer ours.
//
// Without this, disabling a user on the server would leave them able to
// authenticate over SMB with the credential that was just revoked.
func Prune(desired []Entry) ([]string, error) {
	keep := map[string]bool{}
	for _, e := range desired {
		keep[e.Name] = true
	}
	names, err := PassdbNames()
	if err != nil {
		return nil, err
	}
	var removed []string
	for _, name := range names {
		if keep[name] {
			continue
		}
		// A name that failed the rule never reached the database through this
		// agent, and passing one to the tool is what the rule exists to stop.
		if !ValidName(name) {
			continue
		}
		if err := exec.Command("pdbedit", "-x", "-u", name).Run(); err != nil { //nolint:gosec // G204 flags a variable command: the name is held to the portable rule on the line above.
			// One credential that will not delete is reported rather than
			// failing the pass, because the rest still need pruning and a
			// stale one is visible in the next check.
			continue
		}
		removed = append(removed, name)
	}
	return removed, nil
}

// MissingPassdb lists accounts that should be able to authenticate and cannot.
//
// The import tool reports success and imports nothing when a credential line's
// numeric field names no account: no error, no output, a zero exit, an empty
// database. Every downstream symptom points at credentials or configuration
// rather than at the import, so it is checked here where the cause is still
// visible.
func MissingPassdb(desired []Entry) ([]string, error) {
	names, err := PassdbNames()
	if err != nil {
		return nil, err
	}
	known := map[string]bool{}
	for _, n := range names {
		known[n] = true
	}
	var out []string
	for _, e := range desired {
		if !known[e.Name] {
			out = append(out, e.Name)
		}
	}
	return out, nil
}

File: smb-agent/agent/src/accounts.rs
//! `/etc/passwd` reconciliation and the Samba passdb.
//!
//! Samba resolves an SMB login to a Unix account through `getpwnam`, so every
//! account named in a share needs a passwd entry even though `force user`
//! means none of them ever owns a file. `sc-core` renders those entries; this
//! puts them where `getpwnam` looks, and imports the matching NT hashes.
//!
//! Rebuilt from scratch on every pass rather than diffed, which is what makes
//! an account removed from `sc-core`'s registry disappear here too instead of
//! lingering as a stale entry forever.

use std::collections::BTreeSet;
use std::io;
use std::path::Path;
use std::process::Command;

/// Written into the GECOS field of every account this agent creates, and the
/// only thing that makes one eligible for removal. An account without it is
/// somebody else's and is never touched.
const MARKER: &str = "sc-managed-smb";

/// One rendered `passwd` line, split far enough to reason about.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Entry {
    pub name: String,
    pub uid: String,
    pub gid: String,
    /// The whole line, marker already stamped.
    pub line: String,
}

/// Parse what `sc-core` rendered. Anything that is not a seven-field passwd
/// line is dropped rather than passed through: this content ends up in
/// `/etc/passwd`, and a malformed line there breaks `getpwnam` for every
/// account after it.
pub fn parse_rendered(body: &str) -> Vec<Entry> {
    let mut out = Vec::new();
    for line in body.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let mut f: Vec<&str> = line.split(':').collect();
        if f.len() < 7 || f[0].is_empty() {
            continue;
        }
        let (name, uid, gid) = (f[0].to_string(), f[2].to_string(), f[3].to_string());
        f[4] = MARKER;
        out.push(Entry {
            name,
            uid,
            gid,
            line: f.join(":"),
        });
    }
    out
}

/// `sc-core` account names reach `useradd`-shaped territory here, so they are
/// held to the portable-username rule before anything writes them into
/// `/etc/passwd`. A name starting with `-` would also be argument injection
/// into `pdbedit`.
pub fn valid_name(name: &str) -> bool {
    let mut chars = name.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    if !(first.is_ascii_lowercase() || first == '_') {
        return false;
    }
    if name.len() > 32 {
        return false;
    }
    chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '-')
}

/// Whether a passwd line is one of ours.
fn managed(line: &str) -> bool {
    line.split(':').nth(4) == Some(MARKER)
}

/// Refuse to sync at all if an SMB account collides with a pre-existing
/// system account we did not create.
///
/// Adopting one silently would give that account's name an NT hash, and
/// removing it later would delete a system user. The uid half is not
/// cosmetic either: `pdbedit -i` resolves an smbpasswd line to an account by
/// uid and takes whatever name `getpwuid` answers with, so a shared uid
/// attaches the credential to the wrong name.
pub fn collisions(desired: &[Entry], passwd: &str) -> Vec<String> {
    let mut out = Vec::new();
    for e in desired {
        if !valid_name(&e.name) {
            out.push(format!("'{}' is not a valid system account name", e.name));
            continue;
        }
        for line in passwd.lines().filter(|l| !managed(l)) {
            let mut f = line.split(':');
            let (name, _, uid) = (f.next(), f.next(), f.next());
            if name == Some(e.name.as_str()) {
                out.push(format!("'{}' already exists as a non-managed system account", e.name));
            }
            if uid == Some(e.uid.as_str()) {
                out.push(format!(
                    "uid {} is already '{}', a non-managed account: move smb.service_uid clear of the host's real users",
                    e.uid,
                    name.unwrap_or("?")
                ));
            }
        }
    }
    out.sort();
    out.dedup();
    out
}

/// Every group the rendered accounts reference must already exist: this agent
/// does not invent groups, and an account whose gid resolves to nothing
/// breaks `getpwnam` just as surely as a missing account.
pub fn missing_groups(desired: &[Entry], group_file: &str) -> Vec<String> {
    let have: BTreeSet<&str> = group_file
        .lines()
        .filter_map(|l| l.split(':').nth(2))
        .collect();
    let mut out: Vec<String> = desired
        .iter()
        .map(|e| e.gid.clone())
        .filter(|g| !have.contains(g.as_str()))
        .collect();
    out.sort();
    out.dedup();
    out
}

/// The new `/etc/passwd`: everything that is not ours, then ours.
pub fn rebuild(current: &str, desired: &[Entry]) -> String {
    let mut out = String::with_capacity(current.len());
    for line in current.lines().filter(|l| !managed(l)) {
        out.push_str(line);
        out.push('\n');
    }
    for e in desired {
        out.push_str(&e.line);
        out.push('\n');
    }
    out
}

/// Write `/etc/passwd` through a temporary file and a rename, so a reader
/// never sees a partial one. That still leaves the ordinary lost-update race
/// against a concurrent `useradd`; shadow's own lock is not reachable here,
/// and a file server's SMB roster is not a multi-writer passwd.
pub fn write_passwd(path: &Path, body: &str) -> io::Result<()> {
    let tmp = path.with_extension("sc-smb-agent.tmp");
    std::fs::write(&tmp, body)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&tmp, std::fs::Permissions::from_mode(0o644))?;
    }
    std::fs::rename(&tmp, path)
}

/// Import the rendered NT hashes.
pub fn import(smbpasswd: &Path, passdb: &Path) -> io::Result<()> {
    let out = Command::new("pdbedit")
        .arg("-i")
        .arg(format!("smbpasswd:{}", smbpasswd.display()))
        .arg("-e")
        .arg(format!("tdbsam:{}", passdb.display()))
        .output()?;
    if !out.status.success() {
        return Err(io::Error::other(format!(
            "pdbedit import failed: {}",
            String::from_utf8_lossy(&out.stderr).trim()
        )));
    }
    Ok(())
}

/// Accounts the passdb currently holds.
pub fn passdb_names() -> io::Result<Vec<String>> {
    let out = Command::new("pdbedit").arg("-L").output()?;
    if !out.status.success() {
        return Err(io::Error::other("pdbedit -L failed"));
    }
    Ok(String::from_utf8_lossy(&out.stdout)
        .lines()
        .filter_map(|l| l.split(':').next())
        .filter(|n| !n.trim().is_empty())
        .map(|n| n.to_string())
        .collect())
}

/// Drop NT hashes for accounts that are no longer ours. Without this,
/// disabling a user in `sc-core` would leave them able to authenticate over
/// SMB with the credential that was just revoked.
pub fn prune(desired: &[Entry]) -> io::Result<Vec<String>> {
    let keep: BTreeSet<&str> = desired.iter().map(|e| e.name.as_str()).collect();
    let mut removed = Vec::new();
    for name in passdb_names()? {
        if keep.contains(name.as_str()) {
            continue;
        }
        let ok = Command::new("pdbedit")
            .arg("-x")
            .arg("-u")
            .arg(&name)
            .output()
            .map(|o| o.status.success())
            .unwrap_or(false);
        if ok {
            removed.push(name);
        }
    }
    Ok(removed)
}

/// Accounts that should be able to authenticate and cannot.
///
/// `pdbedit -i` reports success and imports nothing when the smbpasswd line's
/// uid field names no passwd entry: no error, no output, exit 0, an empty
/// passdb. Every downstream symptom (a client told the password is wrong,
/// `NT_STATUS_NO_SUCH_USER` in the smbd log) points at credentials or config
/// rather than at the import, so it is checked here where the cause is still
/// visible.
pub fn missing_passdb(desired: &[Entry]) -> io::Result<Vec<String>> {
    let known: BTreeSet<String> = passdb_names()?.into_iter().collect();
    Ok(desired
        .iter()
        .map(|e| e.name.clone())
        .filter(|n| !known.contains(n))
        .collect())
}

#[cfg(test)]
mod tests {
    use super::*;

    const RENDERED: &str = "alice:x:2001:1000::/nonexistent:/sbin/nologin\n\
                            bob:x:2002:1000::/nonexistent:/sbin/nologin\n";

    const SYSTEM: &str = "root:x:0:0:root:/root:/bin/sh\n\
                          scsvc:x:1000:1000::/nonexistent:/sbin/nologin\n";

    #[test]
    fn rendered_entries_get_the_marker() {
        let e = parse_rendered(RENDERED);
        assert_eq!(e.len(), 2);
        assert_eq!(e[0].name, "alice");
        assert_eq!(e[0].uid, "2001");
        assert_eq!(e[0].gid, "1000");
        assert!(e[0].line.contains(MARKER), "{}", e[0].line);
    }

    #[test]
    fn a_malformed_line_never_reaches_etc_passwd() {
        let e = parse_rendered("alice:x:2001\n\nbob:x:2002:1000::/nonexistent:/sbin/nologin\n");
        assert_eq!(e.len(), 1);
        assert_eq!(e[0].name, "bob");
    }

    /// The property the whole marker scheme exists for: an account dropped
    /// from the registry disappears rather than accumulating.
    #[test]
    fn rebuilding_drops_managed_accounts_that_are_gone() {
        let first = rebuild(SYSTEM, &parse_rendered(RENDERED));
        assert!(first.contains("alice"));
        assert!(first.contains("bob"));

        let second = rebuild(&first, &parse_rendered("alice:x:2001:1000::/nonexistent:/sbin/nologin\n"));
        assert!(second.contains("alice"));
        assert!(!second.contains("bob"));
        // And never at the cost of the host's own accounts.
        assert!(second.contains("root:x:0:0"));
        assert!(second.contains("scsvc:x:1000"));
    }

    #[test]
    fn a_name_collision_with_a_real_account_is_refused() {
        let bad = parse_rendered("scsvc:x:2001:1000::/nonexistent:/sbin/nologin\n");
        let c = collisions(&bad, SYSTEM);
        assert!(c.iter().any(|m| m.contains("non-managed system account")), "{c:?}");
    }

    /// The expensive one: `pdbedit -i` resolves by uid, so a shared uid
    /// silently attaches the credential to the wrong name.
    #[test]
    fn a_uid_collision_with_a_real_account_is_refused() {
        let bad = parse_rendered("alice:x:1000:1000::/nonexistent:/sbin/nologin\n");
        let c = collisions(&bad, SYSTEM);
        assert!(c.iter().any(|m| m.contains("uid 1000")), "{c:?}");
    }

    #[test]
    fn our_own_previous_entries_do_not_count_as_collisions() {
        let desired = parse_rendered(RENDERED);
        let passwd = rebuild(SYSTEM, &desired);
        assert!(collisions(&desired, &passwd).is_empty());
    }

    #[test]
    fn names_are_held_to_the_portable_rule() {
        assert!(valid_name("alice"));
        assert!(valid_name("_svc-1"));
        assert!(!valid_name(""));
        assert!(!valid_name("-x"));
        assert!(!valid_name("Alice"));
        assert!(!valid_name("alice bob"));
        assert!(!valid_name(&"a".repeat(33)));
    }

    #[test]
    fn a_gid_with_no_group_is_reported() {
        let g = "root:x:0:\nscsvc:x:1000:\n";
        assert!(missing_groups(&parse_rendered(RENDERED), g).is_empty());
        let orphan = parse_rendered("carol:x:2003:4242::/nonexistent:/sbin/nologin\n");
        assert_eq!(missing_groups(&orphan, g), vec!["4242".to_string()]);
    }

    #[test]
    fn writing_passwd_replaces_it_atomically() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("passwd");
        std::fs::write(&path, SYSTEM).unwrap();
        write_passwd(&path, &rebuild(SYSTEM, &parse_rendered(RENDERED))).unwrap();
        let back = std::fs::read_to_string(&path).unwrap();
        assert!(back.contains("alice"));
        assert!(back.contains("root"));
        assert!(!dir.path().join("passwd.sc-smb-agent.tmp").exists());
    }
}
